### DISCUSSION TITLE
Code Review: Shadowing

### DISCUSSION THREAD

**Discussion by cd155** [JacquesCarette/Drasil#3290] [Show and tell]
In the `isValidParen`, it has a variable name the first `xs`. Inside of let expression, there is a local function `isValidParenHelper`, it has the second `xs`. 

I have two following questions:
1. In this situation, is the first `xs` shadowed by the second `xs`?
2. I intend to change the `x:xs` to `y:ys` to avoid confusion. I just wonder what will you do and what is a good practice?

```haskell
isValidParen :: String -> Bool
isValidParen [] = something
isValidParen xs = 
    let 
        isValidParenHelper :: String -> String -> Bool
        isValidParenHelper [] s = something
        isValidParenHelper (x:xs) s
            | condition1 = isValidParenHelper xs (tail s)
            | otherwise = isValidParenHelper xs (x:s)
    in 
        isValidParenHelper xs []
```

**Comment by JacquesCarette** [comment]
1. yes
2. if you don't want the helper function to have access to the top `xs`, some consider it good practice to shadow it. I'm not so sure.

Personally, I see no reason to make `isValidParenHelper` a local function. I would pull it out into the surrounding module, but then not export it.

**Reply by cd155** [reply]
That's an interesting idea, shadow `xs` when I don't want the helper function to access the top `xs`. Certainly, it makes sense. Shadowing it will erase the previous `xs`, but will it decrease the readability?

One difference would be the `isValidParenHelper` is only accessible by `isValidParen`. If I put it on the surrounding module and even hide it from other modules, any functions in this module can still access it.

**Reply by JacquesCarette** [reply]
Shadowing does reduce "shallow readability", absolutely.
